### PR TITLE
Ability to configure domain_id via InitOptions.

### PR DESCRIPTION
**Pull request by fujitatomoya**
closes https://github.com/ros2/rclcpp/issues/910

Signed-off-by: Tomoya.Fujita <Tomoya.Fujita@sony.com>

### CODE REVIEW

**Comment by fujitatomoya** [comment]
@ivanpauno 

this is frontend fix to have ability to set domain_id in InitOptions, depends on https://github.com/ros2/rcl/pull/678. this needs to be merged after https://github.com/ros2/rcl/pull/678.

File: rclcpp/include/rclcpp/init_options.hpp [hunk 1]
```diff
@@ -80,13 +82,29 @@ class InitOptions
...
 
+  /// Retrieve the ROS_DOMAIN_ID environment variable.
+  RCLCPP_PUBLIC
+  bool
+  use_default_domain_id();
+
+  /// Set domain id.
+  RCLCPP_PUBLIC
+  void
+  set_domain_id(size_t domain_id);
```
[the remark this answers is not in the document]
Comment (fujitatomoya) [reply rclcpp/include/rclcpp/init_options.hpp:93]: i was thinking that way too, we could wait for more feedback on this.

File: rclcpp/include/rclcpp/init_options.hpp [hunk 2]
```diff
@@ -80,13 +82,29 @@ class InitOptions
   const rcl_init_options_t *
   get_rcl_init_options() const;
 
+  /// Retrieve the ROS_DOMAIN_ID environment variable.
+  RCLCPP_PUBLIC
+  bool
```
[the remark this answers is not in the document]
Comment (fujitatomoya) [reply rclcpp/include/rclcpp/init_options.hpp:87]: maybe even w/o exception?

**Comment by fujitatomoya** [comment]
@Barry-Xu-2018 @ivanpauno 

could you take a look at this when you got time? related to https://github.com/ros2/rcl/pull/678. thanks in advance.

**Comment by fujitatomoya** [comment]
@ivanpauno 

friendly ping.

File: rclcpp/src/rclcpp/init_options.cpp [hunk 3]
```diff
@@ -99,4 +99,55 @@ InitOptions::get_rcl_init_options() const
   return init_options_.get();
 }
 
+void
+InitOptions::use_default_domain_id()
+{
+  // Try to get the ROS_DOMAIN_ID environment variable.
+  const char * ros_domain_id = NULL;
```
Comment (ivanpauno) [inline rclcpp/src/rclcpp/init_options.cpp:106]: please reuse https://github.com/ros2/rcl/blob/b9cfc243bf018091b5c5bcb5585d19591daf4ae6/rcl/include/rcl/domain_id.h#L42

File: rclcpp/src/rclcpp/init_options.cpp [hunk 4]
```diff
@@ -99,4 +99,55 @@ InitOptions::get_rcl_init_options() const
...
+  }
+}
+
+void
+InitOptions::set_domain_id(size_t domain_id)
+{
+  rcl_ret_t ret = rcl_init_options_set_domain_id(init_options_.get(), domain_id);
+  if (RCL_RET_OK != ret) {
+    RCLCPP_ERROR(
+      rclcpp::get_logger("rclcpp"), "failed to set domain id (%d).", domain_id);
```
Comment (ivanpauno) [inline rclcpp/src/rclcpp/init_options.cpp:136]: please throw a `RCLError` exception

File: rclcpp/src/rclcpp/init_options.cpp [hunk 5]
```diff
@@ -99,4 +99,55 @@ InitOptions::get_rcl_init_options() const
...
+}
+
+size_t
+InitOptions::get_domain_id() const
+{
+  size_t domain_id;
+  rcl_ret_t ret = rcl_init_options_get_domain_id(init_options_.get(), &domain_id);
+  if (RCL_RET_OK != ret) {
+    RCLCPP_ERROR(
+      rclcpp::get_logger("rclcpp"), "failed to get domain id.");
```
Comment (ivanpauno) [inline rclcpp/src/rclcpp/init_options.cpp:147]: same here

File: rclcpp/src/rclcpp/init_options.cpp (hunk 3 above)
Comment (fujitatomoya) [reply rclcpp/src/rclcpp/init_options.cpp:106]: oversight, fixed.

File: rclcpp/include/rclcpp/init_options.hpp [hunk 6]
```diff
@@ -16,9 +16,11 @@
 #define RCLCPP__INIT_OPTIONS_HPP_
 
 #include <memory>
+#include <limits>
```
Comment (fujitatomoya) [inline rclcpp/include/rclcpp/init_options.hpp:19]: no need to include this header.

File: rclcpp/include/rclcpp/init_options.hpp [hunk 7]
```diff
@@ -16,9 +16,11 @@
...
 
 #include "rcl/init_options.h"
 #include "rclcpp/visibility_control.hpp"
+#include "rcutils/get_env.h"
```
Comment (fujitatomoya) [inline rclcpp/include/rclcpp/init_options.hpp:23]: same here.

File: rclcpp/include/rclcpp/init_options.hpp [hunk 8]
```diff
@@ -80,6 +82,21 @@ class InitOptions
   const rcl_init_options_t *
   get_rcl_init_options() const;
 
+  /// Retrieve the ROS_DOMAIN_ID environment variable.
```
Comment (fujitatomoya) [inline rclcpp/include/rclcpp/init_options.hpp:85]: comment is not correct.

File: rclcpp/src/rclcpp/init_options.cpp [hunk 9]
```diff
@@ -99,4 +99,36 @@ InitOptions::get_rcl_init_options() const
...
+  }
+}
+
+size_t
+InitOptions::get_domain_id() const
+{
+  size_t domain_id;
+  rcl_ret_t ret = rcl_init_options_get_domain_id(init_options_.get(), &domain_id);
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret, "failed to get domain id to rcl init options");
```
Comment (fujitatomoya) [inline rclcpp/src/rclcpp/init_options.cpp:128]:
```suggestion
    rclcpp::exceptions::throw_from_rcl_error(ret, "failed to get domain id from rcl init options");
```

**Comment by fujitatomoya** [comment]
@ivanpauno 

I've addressed to your comments.

File: rclcpp/src/rclcpp/init_options.cpp [hunk 10]
```diff
@@ -99,4 +99,36 @@ InitOptions::get_rcl_init_options() const
   return init_options_.get();
 }
 
+void
+InitOptions::use_default_domain_id()
+{
+  size_t domain_id;
```
Comment (fujitatomoya) [inline rclcpp/src/rclcpp/init_options.cpp:105]: if `ROS_DOMAIN_ID` is not set, uninitialized domain_id will be set.

File: rclcpp/src/rclcpp/init_options.cpp [hunk 11]
```diff
@@ -99,4 +99,36 @@ InitOptions::get_rcl_init_options() const
   return init_options_.get();
 }
 
+void
+InitOptions::use_default_domain_id()
```
Comment (ivanpauno) [inline rclcpp/src/rclcpp/init_options.cpp:103]: It sounds like all setters and getters in this class aren't thread safe.
It would be good to address that problem in another PR, maybe before merging this one

Comment (fujitatomoya) [reply rclcpp/src/rclcpp/init_options.cpp:103]: correct, this needs to be fixed before merging this one.

**Comment by fujitatomoya** [comment]
> Can you add some testing?

will do.

**Comment by fujitatomoya** [comment]
@ivanpauno 

init_options_mutex and test are added, could you take a look when you got time.

thanks in advance,

File: rclcpp/src/rclcpp/init_options.cpp [hunk 12]
```diff
@@ -33,6 +33,7 @@ InitOptions::InitOptions(rcl_allocator_t allocator)
 InitOptions::InitOptions(const rcl_init_options_t & init_options)
 : init_options_(new rcl_init_options_t)
 {
+  std::lock_guard<std::recursive_mutex> init_options_lock(init_options_mutex_);
```
Comment (ivanpauno) [inline rclcpp/src/rclcpp/init_options.cpp:36]: taking a mutex in a constructor doesn't make sense

File: rclcpp/include/rclcpp/init_options.hpp [hunk 13]
```diff
@@ -80,11 +81,28 @@ class InitOptions
...
+  RCLCPP_PUBLIC
+  size_t
+  get_domain_id() const;
+
 protected:
   void
   finalize_init_options();
 
 private:
+  // This mutex is recursive so that the operator can ensure atomicity
```
[the remark this answers is not in the document]
Comment (fujitatomoya) [reply rclcpp/include/rclcpp/init_options.hpp:104]: right, fix it.

File: rclcpp/src/rclcpp/init_options.cpp (hunk 12 above)
Comment (fujitatomoya) [reply rclcpp/src/rclcpp/init_options.cpp:36]: right, fix it.

File: rclcpp/include/rclcpp/init_options.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (ivanpauno) [reply rclcpp/include/rclcpp/init_options.hpp]: > there is a chance that mutex will be locked in InitOptions::finalize_init_options being already locked in assignment operator.

This is only because how the copy assignment operator is implemented.
You can add a private `InitOptions::finalize_init_options_impl` or a free function `finalize_init_options_impl(rcl_init_options_t *)` that doesn't take the mutex.

A `std::recursive_mutex` should only be used when the user can actually call some of the methods that are taking the mutex recursively.

[the remark this answers is not in the document]
Comment (fujitatomoya) [reply rclcpp/include/rclcpp/init_options.hpp]: that makes sense to me, thanks. will fix is as you suggested.

[the remark this answers is not in the document]
Comment (fujitatomoya) [reply rclcpp/include/rclcpp/init_options.hpp]: This has been addressed with https://github.com/ros2/rclcpp/pull/1165/commits/e8fb34ac5bd9ad111d9e9b9a584af12f4cbcf092

**Comment by fujitatomoya** [comment]
@ivanpauno 

done fix, could you take a look once again when you got time?

File: rclcpp/src/rclcpp/init_options.cpp [hunk 14]
```diff
@@ -96,7 +106,42 @@ InitOptions::finalize_init_options()
 const rcl_init_options_t *
 InitOptions::get_rcl_init_options() const
 {
+  std::lock_guard<std::mutex> init_options_lock(init_options_mutex_);
```
Comment (ivanpauno) [inline rclcpp/src/rclcpp/init_options.cpp:109]: Taking the mutex here doesn't help.
The returned pointer can be invalidated any time, as the API signature is unsafe.

For the moment, I would just delete the lock guard.

File: rclcpp/src/rclcpp/init_options.cpp [hunk 15]
```diff
@@ -44,6 +44,7 @@ InitOptions::InitOptions(const InitOptions & other)
 : InitOptions(*other.get_rcl_init_options())
 {
   shutdown_on_sigint = other.shutdown_on_sigint;
+  initialize_logging_ = other.initialize_logging_;
```
Comment (ivanpauno) [inline rclcpp/src/rclcpp/init_options.cpp:47]: I thought I commented this before, but I think I didn't.
I would like to see this fix in a separate PR, so it can be backported.

File: rclcpp/src/rclcpp/init_options.cpp [hunk 16]
```diff
@@ -63,12 +64,14 @@ InitOptions &
...
   if (this != &other) {
-    this->finalize_init_options();
+    std::lock_guard<std::mutex> init_options_lock(init_options_mutex_);
+    this->finalize_init_options_impl();
     rcl_ret_t ret = rcl_init_options_copy(other.get_rcl_init_options(), init_options_.get());
     if (RCL_RET_OK != ret) {
       rclcpp::exceptions::throw_from_rcl_error(ret, "failed to copy rcl init options");
     }
     this->shutdown_on_sigint = other.shutdown_on_sigint;
+    this->initialize_logging_ = other.initialize_logging_;
```
Comment (ivanpauno) [inline rclcpp/src/rclcpp/init_options.cpp:74]: same here, this should be in a separate PR

File: rclcpp/test/rclcpp/test_init_options.cpp [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+  }
+}
+
+TEST(TestInitOptions, test_initialize_logging) {
+  {
+    auto options = rclcpp::InitOptions();
+    EXPECT_TRUE(options.auto_initialize_logging());
+    const rcl_init_options_t * rcl_options = options.get_rcl_init_options();
+    ASSERT_TRUE(rcl_options != nullptr);
+    ASSERT_TRUE(rcl_options->impl != nullptr);
```
Comment (ivanpauno) [inline rclcpp/test/rclcpp/test_init_options.cpp:54]: I don't understand why this three lines have to be repeated here.
This was already tested in the construction test, and it's not related to the `auto_initialize_logging` API.

File: rclcpp/test/rclcpp/test_init_options.cpp [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+  }
+
+  {
+    auto options = rclcpp::InitOptions().auto_initialize_logging(true);
+    EXPECT_TRUE(options.auto_initialize_logging());
+    const rcl_init_options_t * rcl_options = options.get_rcl_init_options();
+    ASSERT_TRUE(rcl_options != nullptr);
+    ASSERT_TRUE(rcl_options->impl != nullptr);
```
Comment (ivanpauno) [inline rclcpp/test/rclcpp/test_init_options.cpp:62]: same here

File: rclcpp/test/rclcpp/test_init_options.cpp [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+  }
+
+  {
+    auto options = rclcpp::InitOptions().auto_initialize_logging(false);
+    EXPECT_FALSE(options.auto_initialize_logging());
+    const rcl_init_options_t * rcl_options = options.get_rcl_init_options();
+    ASSERT_TRUE(rcl_options != nullptr);
+    ASSERT_TRUE(rcl_options->impl != nullptr);
```
Comment (ivanpauno) [inline rclcpp/test/rclcpp/test_init_options.cpp:70]: same here

File: rclcpp/test/rclcpp/test_init_options.cpp [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+  }
+}
+
+TEST(TestInitOptions, test_domain_id) {
+  rcl_allocator_t allocator = rcl_get_default_allocator();
+  auto options = rclcpp::InitOptions(allocator);
+  const rcl_init_options_t * rcl_options = options.get_rcl_init_options();
+  ASSERT_TRUE(rcl_options != nullptr);
+  ASSERT_TRUE(rcl_options->impl != nullptr);
```
Comment (ivanpauno) [inline rclcpp/test/rclcpp/test_init_options.cpp:79]: same here

File: rclcpp/test/rclcpp/test_init_options.cpp (hunk 17 above)
Comment (fujitatomoya) [reply rclcpp/test/rclcpp/test_init_options.cpp:54]: right, i was thinking that `just in case`, but as you mentioned, It'd be better to keep the consistency.

**Comment by fujitatomoya** [comment]
> I would like to see the bug being fixed in an independent PR, so it can be easily backported.

that makes sense, i will delete unrelated minor fixes from this PR. (I will do that after this PR is merged :smile:)

**Comment by fujitatomoya** [comment]
@ivanpauno 

appreciate for checking on this, could you take a look once again? btw, i did delete anything related to `initialize_logging_` including test since test does not pass w/o fix. so i will make another PR for `initialize_logging_` with test.

**Comment by ivanpauno** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11755/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6832/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9590/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11698/)

**Comment by ivanpauno** [comment]
@fujitatomoya can you check the failures?

**Comment by fujitatomoya** [comment]
I cannot reproduce this failure in local environment with `colcon test`,

> https://ci.ros2.org/job/ci_linux/11757/testReport/projectroot/test/test_init_options/

as far as i can tell, ROS_DOMAIN_ID environmental variable is set in CI. So test needs to check if it is set before calling `InitOptions::use_default_domain_id` and compare it against after.

**Comment by fujitatomoya** [comment]
@ivanpauno 

https://github.com/ros2/rclcpp/pull/1165/commits/300580356c5bb665d999b0c6861197dc945944c1 should be able to fix the CI.

**Comment by ivanpauno** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11767/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6844/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9604/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11709/)

**Comment by ivanpauno** [comment]
Thanks for all the iterations on the review @fujitatomoya !

**Comment by ivanpauno** [comment]
@fujitatomoya If you have the time, it would also be nice to add a `get_domain_id` method to `Context`.
That should wrap `rcl_context_get_domain_id`

**Comment by fujitatomoya** [comment]
@ivanpauno 

> If you have the time, it would also be nice to add a get_domain_id method to Context.

will do 😄
